pkg/ccl/crosscluster/logical/lww_row_processor_test.go

```diff
@@ -301,3 +306,170 @@ func BenchmarkLWWInsertBatch(b *testing.B) {
+			"ALTER TABLE %s "+lwwColumnAdd,
+			tableName))
+		tableNumber++
+		return tableName
+	}
+
+	// The encoderFn takes an origin timestamp and a row and converts into a Key-Value format that
+	// can be ingested by the RowProcessor
+	type encoderFn func(originTimestamp hlc.Timestamp, datums ...interface{}) roachpb.KeyValue
+
+	setup := func(t *testing.T, useKVProc bool) (string, RowProcessor, encoderFn) {
+		tableNameSrc := createTable(t)
+		tableNameDst := createTable(t)
+		srcDesc := desctestutils.TestingGetPublicTableDescriptor(s.DB(), s.Codec(), "defaultdb", tableNameSrc)
+		dstDesc := desctestutils.TestingGetPublicTableDescriptor(s.DB(), s.Codec(), "defaultdb", tableNameDst)
+
+		// We need the SQL row processor even when testing the KW row processor since it's the fallback
+		var rp RowProcessor
+		rp, err := makeSQLProcessor(ctx, s.ClusterSettings(), map[descpb.ID]sqlProcessorTableConfig{
+			dstDesc.GetID(): {
+				srcDesc: srcDesc,
+			},
+		}, jobspb.JobID(1), s.InternalExecutor().(isql.Executor))
+		require.NoError(t, err)
+
+		if useKVProc {
+			rp, err = newKVRowProcessor(ctx,
+				&execinfra.ServerConfig{
+					DB:           s.InternalDB().(descs.DB),
+					LeaseManager: s.LeaseManager(),
+				}, &eval.Context{
+					Codec:    s.Codec(),
+					Settings: s.ClusterSettings(),
+				}, map[descpb.ID]sqlProcessorTableConfig{
+					dstDesc.GetID(): {
+						srcDesc: srcDesc,
+					},
+				},
+				rp.(*sqlRowProcessor))
+			require.NoError(t, err)
+		}
+		return tableNameDst, rp, func(originTimestamp hlc.Timestamp, datums ...interface{}) roachpb.KeyValue {
+			kv := replicationtestutils.EncodeKV(t, s.Codec(), srcDesc, datums...)
+			kv.Value.Timestamp = originTimestamp
+			return kv
+		}
+	}
+
+	insertRow := func(rp RowProcessor, keyValue roachpb.KeyValue, prevValue roachpb.Value) error {
+		return s.InternalDB().(isql.DB).Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			_, err := rp.ProcessRow(ctx, txn, keyValue, prevValue)
+			return err
+		})
+	}
+
+	timeNow := hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+	timeNowPlusOne := hlc.Timestamp{WallTime: timeutil.Now().Add(time.Microsecond).UnixNano()}
+	timeOneDayForward := hlc.Timestamp{WallTime: timeutil.Now().Add(time.Hour * 24).UnixNano()}
+	timeOneDayBackward := hlc.Timestamp{WallTime: timeutil.Now().Add(time.Hour * -24).UnixNano()}
+	row1 := []interface{}{1, "row1"}
+	row2 := []interface{}{1, "row2"}
+	row3 := []interface{}{1, "row3"}
+
+	// Run with both the SQL row processor and the KV row processor. Note that currently,
+	// KV row proc write conflicts result in a fallback to the SQL row processor.
+	testutils.RunTrueAndFalse(t, "useKVProc", func(t *testing.T, useKVProc bool) {
+		// All of those combinations are tested with both optimistic inserts and standard inserts.
+		testutils.RunTrueAndFalse(t, "optimistic_insert", func(t *testing.T, optimisticInsert bool) {
+
+			// Write to table on both the src and dest cluster and see how conflicts are handled
```

Comment:
i think this doctring could be sharper: "When a remote insert conflicts with a local write, comparing the origin mvcc timestamp of the remote write to the local write's mvcc timestamp determines LWW".